MappingSchema _normalize can raise unclear "StopIteration" error if supplied schema dict has table with no columns

I hit this while writing unit tests trying to reproduce an unrelated issue related to qualifying snowflake stages, but figured I'd write it up here - After digging through SQLGlot code, I found a spot where the error handling could possible be improved for future library users.

**Fully reproducible code snippet**
 

The below raises a confusing `StopIteration` error

```python
catalog_name = "test_catalog"
database_name = "test_db"
schema = {
    "test_catalog": {
        "test_db": {
            "linear_labels_8589937695": {
            },
        }
    }
}

query_text = """
  CREATE OR REPLACE TABLE LINEAR_LABELS_8589937695 (
    _PORTABLE_EXTRACTED TIMESTAMP,
  ) AS SELECT 
      s.$1:_PORTABLE_EXTRACTED,
  FROM @portable_stage_38654707232/linear_labels_85899376951796365056 s;
"""

parsed = parse_one(query_text, dialect="snowflake")
qualified = qualify(
    parsed,
    db=database_name,
    catalog=catalog_name,
    schema=schema,
    dialect="snowflake",
    infer_schema=True,
    expand_stars=False, 
    validate_qualify_columns=False,  
    qualify_columns=False, 
)
```

<details><summary>Full Error Trace</summary>
    
    ../odp/core/detect_unused.py:143: in extract_tables
        qualified = qualify(
    ../examples/snowflake-dagster-asset/venv/lib/python3.11/site-packages/sqlglot/optimizer/qualify.py:67: in qualify
        schema = ensure_schema(schema, dialect=dialect)
    ../examples/snowflake-dagster-asset/venv/lib/python3.11/site-packages/sqlglot/schema.py:478: in ensure_schema
        return MappingSchema(schema, **kwargs)
    ../examples/snowflake-dagster-asset/venv/lib/python3.11/site-packages/sqlglot/schema.py:232: in __init__
        super().__init__(self._normalize(schema) if self.normalize else schema)
    ../examples/snowflake-dagster-asset/venv/lib/python3.11/site-packages/sqlglot/schema.py:373: in _normalize
        if isinstance(first(columns.values()), dict):
    _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
    
    it = dict_values([])
    
        def first(it: t.Iterable[T]) -> T:
            """Returns the first element from an iterable (useful for sets)."""
    >       return next(i for i in it)
    E       StopIteration

</details>

It looks like this is coming from here https://github.com/tobymao/sqlglot/blob/main/sqlglot/schema.py#L389, where the buildtin `first()` is called on an empty dict.

Obviously my usage here is not correct, but it took me some digging through the code to figure that out, because the `StopIteration` error wasn't very clear. It seems that in the spirit of the other validation/checking in that method, we'd want to raise a similar `SchemaError` if the table in the dict has no columns for whatever reason, perhaps something like


```diff
-            if not isinstance(columns, dict):
+            if not isinstance(columns, dict) or len(columns) == 0:
                raise SchemaError(error_msg.format(".".join(keys[:-1]), len(flattened_schema[0])))
            if isinstance(first(columns.values()), dict):
                raise SchemaError(
                    error_msg.format(
                        ".".join(keys + flatten_schema(columns)[0]), len(flattened_schema[0])
                    ),
                )
```

--- a/sqlglot/schema.py
+++ b/sqlglot/schema.py
@@ -386,6 +386,8 @@ class MappingSchema(AbstractMappingSchema, Schema):
 
             if not isinstance(columns, dict):
                 raise SchemaError(error_msg.format(".".join(keys[:-1]), len(flattened_schema[0])))
+            if not columns:
+                raise SchemaError(f"Table {'.'.join(keys[:-1])} must have at least one column")
             if isinstance(first(columns.values()), dict):
                 raise SchemaError(
                     error_msg.format(

